cmd/lotus-chainwatch/sync.go

```diff
@@ -263,57 +305,97 @@ func syncHead(ctx context.Context, api api.FullNode, st *storage, headTs *types.
 
 		log.Infof("Getting miner info")
 
-		miners := map[minerKey]*minerInfo{}
+		// map of tipset to all miners that had a head-change at that tipset.
+		minerTips := map[types.TipSetKey][]*newMinerInfo{}
+		// heads we've seen, im being paranoid
+		headsSeen := map[cid.Cid]bool{}
 
+		minerChanges := 0
 		for addr, m := range actors {
 			for actor, c := range m {
 				if actor.Code != builtin.StorageMinerActorCodeID {
 					continue
 				}
 
-				miners[minerKey{
+				// only want miner actors with head change events
+				if headsSeen[actor.Head] {
+					continue
+				}
+				minerChanges++
+
+				minerTips[c.tsKey] = append(minerTips[c.tsKey], &newMinerInfo{
 					addr:      addr,
 					act:       actor,
 					stateroot: c.stateroot,
-					tsKey:     c.tsKey,
-				}] = &minerInfo{}
-			}
-		}
 
-		parmap.Par(50, parmap.KVMapArr(miners), func(it func() (minerKey, *minerInfo)) {
-			k, info := it()
+					state: miner.State{},
+					info:  miner.MinerInfo{},
 
-			// TODO: get the storage power actors state and and pull the miner power from there, currently this hits the
-			// storage power actor once for each miner for each tipset, we can do better by just getting it for each tipset
-			// and reading each miner power from the result.
-			pow, err := api.StateMinerPower(ctx, k.addr, k.tsKey)
-			if err != nil {
-				log.Error(err)
-				// Not sure why this would fail, but its probably worth continuing
-			}
-			info.power = pow.MinerPower.QualityAdjPower
+					rawPower: big.Zero(),
+					qalPower: big.Zero(),
+					ssize:    0,
+					psize:    0,
+				})
 
-			sszs, err := api.StateMinerSectorCount(ctx, k.addr, k.tsKey)
-			if err != nil {
-				log.Error(err)
-				return
+				headsSeen[actor.Head] = true
 			}
-			info.psize = sszs.Pset
-			info.ssize = sszs.Sset
+		}
+
+		minerProcessingState := time.Now()
```

Comment:
Nit: Slightly more descriptive var name. 🙏 
```suggestion
		minerProcessingStartedAt := time.Now()
```